### PR TITLE
Feat: orphanedRowAction=disabled (rebase of PR 8285)

### PR DESCRIPTION
**Pull request by jbjhjm**
@pleerock 

This is a new PR based on orphanedRowAction=skip PR https://github.com/typeorm/typeorm/pull/8285

Fixes https://github.com/typeorm/typeorm/issues/8275

Implementation is as before, just rebased to include commits added in the meantime.
Also, instead of keyword "skip" it is now using "disabled", as discussed in other PR.

### CODE REVIEW

File: docs/relations.md [hunk 1]
```diff
@@ -25,7 +25,8 @@ There are several options you can specify for relations:
 * `onDelete: "RESTRICT"|"CASCADE"|"SET NULL"` - specifies how foreign key should behave when referenced object is deleted
 * `primary: boolean` - Indicates whether this relation's column will be a primary column or not.
 * `nullable: boolean` - Indicates whether this relation's column is nullable or not. By default it is nullable.
-* `orphanedRowAction: "nullify" | "delete" | "soft-delete"` - When a child row is removed from its parent, determines if the child row should be orphaned (default) or deleted (delete or soft delete).
+* `orphanedRowAction: "nullify" | "delete" | "soft-delete" | disable` - When a parent is saved (cascading enabled) without a child/children that still exists in database, this will control what shall happen to them.
```
Comment (pleerock) [inline docs/relations.md:28]: take new option into quotes

**Comment by pleerock** [comment]
Tests are failing....

**Comment by jbjhjm** [comment]
@pleerock postgres test fails because of unexpected data, but this must be a quirk within relation update mechanism.
As you can see in PR file diff, the change only touches handling of removedRelatedEntitySubject.
Should I just simplify the test so that it passes with postgres too?

Shortened info on what is happening:

```ts
const userToUpdate = (await userRepo.findOne(userId))!;
userToUpdate.settings = [
  userToUpdate.settings[0],
  // create modified version of setting at index 1
  { ...userToUpdate.settings[1], data:"bar_updated" },
];
await userRepo.save(userToUpdate);
```

then the user is fetched again and its settings entries are checked.
(only) on postgres, userToUpdate.settings[1] will not have been updated.

File: docs/relations.md (hunk 1 above)
Comment (pleerock) [reply docs/relations.md:28]: `disable` must be `"disable"`

**Comment by pleerock** [comment]
@jbjhjm  not sure I understand the problem

**Comment by jbjhjm** [comment]
> @jbjhjm not sure I understand the problem

Postgres test fails because it didn't persist changes on related item correctly.
Problem seems unrelated to this PR as it only changes handling of removedRelatedEntitySubject.
I can take another look but I have no idea about postgres and its quirks.

**Comment by pleerock** [comment]
Also it looks weird that test is only failing for postgres version 12

**Comment by jbjhjm** [comment]
Oh, I was not aware of 0.3.0 @pleerock ! Let's continue with 0.2.x but I'll also open a mirrored PR for the 0.3.0 branch, ok?

**Comment by ivansieder** [comment]
I've been digging into some issues related to unwanted nullified references for quiet some hours until I came across this issue :D

just for reference, I'm adding another use-case for this PR (thank you @jbjhjm in the meantime for your effort you've put into this PR). We have soft-deletion in place (a custom one due to how other applications handled it in the past) and when updating one-to-many relations and not passing some soft-deleted children, it tries to set it's reference to null, even if it should just ignore it.

So looking really forward to this pull request :)

**Comment by pleerock** [comment]
@jbjhjm sorry to annoy you again, but I would like to ask you to rebase it against latest master, since latest master is 0.3.0 already, and I don't have plans to accept new PRs into 0.2.0

**Comment by jbjhjm** [comment]
@pleerock will do but it'll take some time, haven't even touched 0.3.x yet due to the breaking changes and tight time.

**Comment by pleerock** [comment]
@jbjhjm np

**Comment by jbjhjm** [comment]
@pleerock alright, I ported it to current master. I was not able to run any local tests, getting a Jscript error popup stating typeorm cli.js contains an error. No idea. Are tests currently run on CircleCI? I see only lint + build checks here.

**Comment by pleerock** [comment]
you need to run prettier to fix lint issues. Also please merge latest master into your branch again because it contains few important fixes on tests are running in CI.

**Comment by jbjhjm** [comment]
fixed, looks good @pleerock

**Comment by jbjhjm** [comment]
@pleerock @AlexMesser please check. I did half a dozen rebases for this due to waiting time and I really don't want to do it another time.

**Comment by pleerock** [comment]
thanks you for all your efforts and patience @jbjhjm

**Comment by nrathi** [comment]
@jbjhjm sorry to bother you, I'm having trouble understanding this option you added.

I have a parent & child entity (renamed to "Document" and "Page" for this example). I want to ensure the following:

1. A page is never an orphan -- it's always associated with a document, and if something would jeopardize that, an error is thrown. Is this correctly accomplished by orphanedRowAction: "disable"?
2. The pages belonging to a document must be unique in page_number.  Is this correctly accomplished by @Unique("document_page_number", ["document", "page_number"])?

```
@Entity("document")
export class Document {
  @OneToMany(() => Page, (listing) => listing.pages, {
    eager: true,
    cascade: true,
  })
  pages!: Page[];

@Entity("page")
@Unique("document_page_number", ["document", "page_number"])
export class MarketplaceListing extends Model<InternalMarketplaceListingId> {
  @ManyToOne(() => Document, (item) => item.pages, {
    nullable: false,
    orphanedRowAction: "disable",
  })
  document!: Document;

  @Column()
  page_number!: number;
}
```

**Comment by nrathi** [comment]
@jbjhjm sorry to bother you, I'm having trouble understanding this option you added.

I have a parent & child entity (renamed to "Document" and "Page" for this example). I want to ensure the following:

1. A page is never an orphan -- it's always associated with a document, and if something would jeopardize that, an error is thrown. Is this correctly accomplished by `orphanedRowAction: "disable"`?
2. The pages belonging to a document must be unique in page_number.  Is this correctly accomplished by `@Unique("document_page_number", ["document", "page_number"])`?

```
@Entity("document")
export class Document {
  @OneToMany(() => Page, (listing) => listing.pages, {
    eager: true,
    cascade: true,
  })
  pages!: Page[];

@Entity("page")
@Unique("document_page_number", ["document", "page_number"])
export class Page  {
  @ManyToOne(() => Document, (item) => item.pages, {
    nullable: false,
    orphanedRowAction: "disable",
  })
  document!: Document;

  @Column()
  page_number!: number;
}
```

**Comment by jbjhjm** [comment]
Hi @nrathi , I think orphanedRowAction is not what you are looking for.
It controls typeOrm's behavior when SAVING an entity which contains a OneToMany property.

To prevent deletion of a Document that still has Pages assigned, take a look at the "onDelete" option. 
It depends on native database features however, so it may work differently in your project than in mine (mysql/mariaDb).
I suggest to set onDelete = "NO ACTION". This will make the database driver throw an error when trying to delete a row which still  has references in other table rows (foreign key constraints.)

Also take a look at this: https://github.com/typeorm/typeorm/issues/8277
In that issue, I summed up all documentation, hints and my own experiences all around configuring Relations.

About the `@Unique` setup, I don't have helpful knowledge about that.

Good luck!

**Pull request merged**